### PR TITLE
client: fix the pd client could be blocked in some cases

### PR DESCRIPTION
**Pull request by leoppro**
Signed-off-by: leoppro <zhaoyilin@pingcap.com>

### What problem does this PR solve?

Fix the PD client could be blocked in some cases

### What is changed and how it works?

There are two levels of context in PD client: `client context` and `tso request context`.
the `tsoRequest.Wait()` function could be blocked when the `client context` is canceled because the `Wait()` function haven't listened the `client context`

### Benchmark
the version of PD server: v4.0.8

before this PR:

```
Start benchmark #0, duration: 5s
Create 3 client(s) for benchmark
count:768290, max:9, min:0, avg:0, >1ms:34372, >2ms:2037, >5ms:457, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:800952, max:9, min:0, avg:0, >1ms:96750, >2ms:13593, >5ms:1495, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:745906, max:8, min:0, avg:0, >1ms:12877, >2ms:3652, >5ms:1330, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:759058, max:4, min:0, avg:0, >1ms:11232, >2ms:5776, >5ms:0, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0

Total:
count:3074206, max:9, min:0, avg:0, >1ms:155231, >2ms:25058, >5ms:3282, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:3074206, >1ms:5.05%, >2ms:0.82%, >5ms:0.11%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%
```

after this PR:

```
Start benchmark #0, duration: 5s
Create 3 client(s) for benchmark
count:1361814, max:7, min:0, avg:1, >1ms:900605, >2ms:204456, >5ms:2896, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:1357623, max:11, min:0, avg:1, >1ms:931611, >2ms:151281, >5ms:3936, >10ms:1, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:737217, max:7, min:0, avg:0, >1ms:3830, >2ms:3444, >5ms:1028, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:1033124, max:8, min:0, avg:0, >1ms:430725, >2ms:83313, >5ms:460, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0

Total:
count:4489778, max:11, min:0, avg:1, >1ms:2266771, >2ms:442494, >5ms:8320, >10ms:1, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:4489778, >1ms:50.49%, >2ms:9.86%, >5ms:0.19%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%
```

### Check List

Tests

- Unit test

### Release note

- Fix the pd client could be blocked in some cases

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -140,6 +140,7 @@ func WithRetry(retry uint64) RegionsOption {
 
 type tsoRequest struct {
 	start      time.Time
+	client     *client
```
Comment (JmPotato) [inline client/client.go:143]: Maybe we can just pass the `client.ctx` here rather than the whole `client`? It's more specific and clear.

**Review by overvenus — APPROVED** [review]
LGTM, could you benchmark the change with`tools/pd-tso-bench`?

**Review by JmPotato — APPROVED** [review]
LGTM

**Comment by JmPotato** [comment]
@leoppro Could you please bring this change to `release-4.0` also? The PD client in that branch should have the same problem.

**Comment by leoppro** [comment]
> @leoppro Could you please bring this change to `release-4.0` also? The PD client in that branch should have the same problem.

ok

**Comment by HunDunDM** [comment]
I think it is not appropriate to use multiple ctx, I think this PR needs more discussion.

**Comment by Yisaer** [comment]
> There are two levels of context in PD client: client context and tso request context.

I think when `clientCtx` is closed, the caller should guranteed the `requestCtx` should be closed too. That is to say, it's a bug or improvement in pdClient. @leoppro @overvenus

**Comment by leoppro** [comment]
> > There are two levels of context in PD client: client context and tso request context.
> 
> I think when `clientCtx` is closed, the caller should guarantee the `requestCtx` should be closed too. That is to say, it's not a bug or improvement in pdClient. @leoppro @overvenus

I don't agree with you. Let's think about this following cases:

```
client:=pd.NewClient(xxx) // this line is equivalent to `client:=pd.NewClientWithContext(context.Backend(),xxx)`

// goroutine1
client.GetTSO(context.Backend(), xx)

// goroutine2 
client.Close() // this will lead to client context canceled, and goroutine1 will be blocked
```

this case is exist in tikv client:

here we call GetTSO using backend context: https://github.com/pingcap/tidb/blob/master/store/tikv/kv.go#L387
and here we create pd client using backend context: https://github.com/pingcap/tidb/blob/master/store/tikv/kv.go#L88
note that I can't specify a context for this as a user of tikv client.

**Comment by HunDunDM** [comment]
I found a similar double `ctx` design in [grpc](https://github.com/grpc/grpc-go/blob/v1.26.0/clientconn.go#L530), and I think it can be merged.

**Pull request merged**